### PR TITLE
Use Pillar to choose which interfaces to monitor.

### PR DESCRIPTION
**Pull request by jm890**
The `linux_netlink.ls` function used a regex to choose which interfaces
to collect metric for.

`_alphanum_re = re.compile(r'^[a-z0-9]+$')`

Unfortunately, by default this excludes vlan and tap interfaces, which
are kind of important.  ie `bond0.120` or `tap2a3dab86-fb`.

We also have a problem where even if we update the regex to include
these interfaces... if someone deletes and spawns a new instance then
the tap device name changes on the compute host, which will not be
monitored unless someone re-runs the `collectd` on the compute again.
Less than ideal.

This commit lets us choose `VerboseInterface "all"` using Pillar data
to avoid this problem.

### CODE REVIEW

**Comment by cloudlab-autobuild** [comment]
Can one of the admins verify this patch?

**Comment by simonpasquier** [comment]
I have some concerns with this change:
* Monitoring `tap` interfaces will increase the number of series in InfluxDB which is something we want to minimize. Especially since you kind of get the same metrics from the libvirt plugin for compute.
* Changing the default value to `VerboseInterface "all"` will impact all existing deployments.

Instead I would rather keep the `linux_netlink.ls` module and extend the Pillar structure to be able to customize the regex pattern. And you could say that when the parameter is empty, it means `VerboseInterface "all"` if really there is a need for it.

**Comment by epcim** [comment]
[**@simonpasquier**](https://github.com/simonpasquier) commented on [Jun 15, 2017, 9:02 AM GMT+2](https://github.com/salt-formulas/salt-formula-linux/pull/113#issuecomment-308647138 "2017-06-15T07:02:49Z - Replied by Github Reply Comments"):
> I have some concerns with this change:
> * Monitoring `tap` interfaces will increase the number of series in InfluxDB which is something we want to minimize. Especially since you kind of get the same metrics from the libvirt plugin for compute.
> * Changing the default value to `VerboseInterface "all"` will impact all existing deployments.
> 
> Instead I would rather keep the `linux_netlink.ls` module and extend the Pillar structure to be able to customize the regex pattern. And you could say that when the parameter is empty, it means `VerboseInterface "all"` if really there is a need for it.

agree, rather customize regex pattern

**Comment by jm890** [comment]
I still don't see much value in not collecting all interfaces, especially since the libvirt plugin doesn't provide the number of extra metrics that netlink does.  But I can't argue with not changing default behavior.  I'll submit a new commit in a bit.

**Comment by jm890** [comment]
Ok, I've updated it to preserve the regex as the default, but allow overriding either by manually listing interfaces (or "all"), or the regex.

File: _modules/linux_netlink.py [hunk 1]
```diff
@@ -2,16 +2,15 @@
...
-_alphanum_re = re.compile(r'^[a-z0-9]+$')
-_lo_re = re.compile(r'^lo$')
 
-
-def _filter(interface):
-    return _alphanum_re.match(interface) and not _lo_re.match(interface)
-
-
-def ls():
+def ls(regex=r'^[a-z0-9]+$'):
```
Comment (simonpasquier) [inline _modules/linux_netlink.py]: no need for a default value since it's already set in map.jinja

Comment (jm890) [inline _modules/linux_netlink.py]: Ah you are right.

File: linux/meta/collectd.yml [hunk 2]
```diff
@@ -1,11 +1,17 @@
+{%- from "linux/map.jinja" import monitoring with context %}
 local_plugin:
   linux_network_netlink:
     plugin: netlink
     template: linux/files/collectd_netlink.conf
-    ignore_selected: false
-    {%- if 'linux_netlink.ls' in salt.keys() %}
+    ignore_selected: {{ monitoring.netlink.ignore_selected }}
+    {%- if monitoring.netlink.interface_override and monitoring.netlink.interfaces is list %}
```
Comment (cznewt) [inline linux/meta/collectd.yml]: i'd rather test is list and lenght of list > 0 instead of anothar param [interface_override]

Comment (jm890) [inline linux/meta/collectd.yml]: Updated

**Pull request merged**